### ISSUE TITLE
ERROR: Connection refused

### ISSUE THREAD

**Issue by marc-** [chef/chef-provisioning-docker#23]
Getting following error on `machine_create` while run chef-client in local mode:
```
$ sudo CHEF_DRIVER=docker bundle exec chef-client -z recipes/robot.rb
  <....>
    * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
    
    [2015-02-19T10:31:15+00:00] INFO: *** Chef 12.0.3 ***
    [2015-02-19T10:31:15+00:00] INFO: Chef-client pid: 10
    [2015-02-19T10:31:16+00:00] ERROR: Connection refused connecting to http://localhost:8889/nodes/robot, retry 1/5
    [2015-02-19T10:31:21+00:00] ERROR: Connection refused connecting to http://localhost:8889/nodes/robot, retry 2/5
    [2015-02-19T10:31:26+00:00] ERROR: Connection refused connecting to http://localhost:8889/nodes/robot, retry 3/5
    [2015-02-19T10:31:31+00:00] ERROR: Connection refused connecting to http://localhost:8889/nodes/robot, retry 4/5
    [2015-02-19T10:31:36+00:00] ERROR: Connection refused connecting to http://localhost:8889/nodes/robot, retry 5/5
    
    ================================================================================
    Chef encountered an error attempting to load the node data for "robot"
    ================================================================================
    
    Networking Error:
    -----------------
    Connection refused - Connection refused connecting to http://localhost:8889/nodes/robot, giving up
    
    Your chef_server_url may be misconfigured, or the network could be down.
    
    Relevant Config Settings:
    -------------------------
    chef_server_url  "http://localhost:8889"
<...>
```

Here is how `recipes/robot.rb` looks:
```ruby
require 'chef/provisioning/docker_driver'
with_driver 'docker'
machine_image 'robot' do
  recipe 'robot'
  machine_options :docker_options => {
      :base_image => {
          :name => 'ubuntu',
          :repository => 'ubuntu',
          :tag => '14.04'
      }
  }
end
```

Obviously this happens due to chef-zero is running on host and container has no access to it. This could be solved by adding `--net=host` option to container when running chef-client in local mode. Another way is probably to play with iptables.

This issue also occures when testing recipes with test-kitchen.

**Comment by scottymarshall** [comment]
@marc- Do you have a work around in place, or are did you just stop testing with test-kitchen and docker for now?

**Comment by jkeiser** [comment]
This indicates something wrong with (or unimplemented with) the transport's proxy.

**Comment by marc-** [comment]
@scottymarshall, didn't find any workaroud so far. Basicaly this issue came from test-kitchen. Was running it with vagrant driver, which game same error. Here is my .kitchen.yml file:
```yaml
---
driver:
  name: vagrant
driver_config:
  use_vagrant_berkshelf_plugin: true
  use_vagrant_provision: true
  require_chef_omnibus: 11.18

provisioner:
  name: chef_zero

platforms:
  - name: ubuntu-14.04-s1
    driver:
        box: local-ubuntu-14.04-s1
        box_url: file:///home/user/vagrant/ubuntu-1404-s1.box
  - name: centos-7.0-s1
    driver:
        box: local-centos-7.0-s1
        box_url: file:///home/user/vagrant/centos-70-s1.box

suites:
  - name: default
    cookbooks_path: ../../cookbooks
    data_bags_path: ../../data_bags
    encrypted_data_bags_path: ../../data_bags
    roles: ../../roles
    run_list:
      - recipe[robot]
```

**Comment by marc-** [comment]
@jkeiser, I was looking at the chef_zero_http_proxy.rb, but it's not clear for me at the moment how it supposed to work, sinse chef-zero listening on localhost of the docker host and container has no access to it (if --net=host parameter no set). I was trying to play with iptables rules, it seems not possible to route container's requiests to 127.0.0.1 .

**Comment by scottymarshall** [comment]
@marc- I did find a work around for now, but I am not yet confident that I would break anything else if I submitted it as a pull request:
in `lib/chef/provisioning/docker_driver/docker_transport.rb`:
**Line 64**:
`args = ['docker', 'run', '--name', container_name]`
change to:
`args = ['docker', 'run', '--net', 'host', '--name', container_name]`

**Comment by marc-** [comment]
@scottymarshall, @jkeiser, I found what was wrong with the proxy:
after assignment (file `docker_transport.rb`)
```ruby
L175: host = Socket.getaddrinfo(uri.host, uri.scheme, nil, :STREAM)[0][3]
```
`host` gets '::1' value
So condition
```ruby
L178: if host == '127.0.0.1' || host == '[::1]'
```
is never satisfied. Changing `host == '[::1]'` to host == '::1' resolved issue in my case.

I works on Ubuntu, however, didn't manage it to work on CentOS 7.0 . Still investigating.

@scottymarshall, 
I think that there should be ability to pass docker options as `machine_image` arguments. It shouldn't be hardcoded into transport.

**Comment by jkeiser** [comment]
@marc- oh nice!  I'm unclear on when [::1] vs. ::1 will be right; but adding a third clause for simple '::1' may suffice!

**Comment by marc-** [comment]
@jkeiser, could `[::1]` be wrong from the very begining? Is there a test against it? I know that this is valid notation for http, but I think we need confirmation that `Socket.getaddrinfo(uri.host, uri.scheme, nil, :STREAM)[0][3]` also can return `[::1]`.

**Comment by marc-** [comment]
I was trying following test on Linux da9de053468d 3.16-3-amd64 #1 SMP Debian 3.16.5-1 (2014-10-10) x86_64 GNU/Linux , ruby 1.9.3, 2.1.5, 2.2.0:
```ruby
require 'socket'
hosts = Socket.getaddrinfo('::1', 'http', nil, :STREAM)
puts hosts.inspect
hosts = Socket.getaddrinfo('::1', 'https', nil, :STREAM)
puts hosts.inspect
hosts = Socket.getaddrinfo('localhost', 'http', nil, :STREAM)
puts hosts.inspect
hosts = Socket.getaddrinfo('localhost', 'https', nil, :STREAM)
puts hosts.inspect
```
which always gave me 
```
[["AF_INET6", 80, "::1", "::1", 10, 1, 6]]
[["AF_INET6", 443, "::1", "::1", 10, 1, 6]]
[["AF_INET6", 80, "::1", "::1", 10, 1, 6], ["AF_INET", 80, "127.0.0.1", "127.0.0.1", 2, 1, 6]]
[["AF_INET6", 443, "::1", "::1", 10, 1, 6], ["AF_INET", 443, "127.0.0.1", "127.0.0.1", 2, 1, 6]]
```
Can someone do the same with boot2docker on Windows or Mac?

**Comment by jkeiser** [comment]
@marc- I suspect you are right.  Searching other codebases, the [] syntax seems to be a URL thing.

**Comment by jkeiser** [comment]
This also matches the other provisioners (they don't do [::1]).  I'm calling it a typo and merging.  Thanks soooo much!

**Comment by jkeiser** [comment]
Merged 168db4cc541b955fe2e8fd8d3c67cc14fb92db31 to master.
